cryptographic_estimators/UOVEstimator/UOVAlgorithms/kipnis_shamir.py

```diff
@@ -0,0 +1,116 @@
+# ****************************************************************************
+# Copyright 2023 Technology Innovation Institute
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+# ****************************************************************************
+
+
+from ..uov_algorithm import UOVAlgorithm
+from ..uov_problem import UOVProblem
+from math import log2
+from sage.arith.misc import is_power_of_two
+
+
+class KipnisShamir(UOVAlgorithm):
+    """
+    Construct an instance of Kipnis-Shamir estimator
+
+    in []_, Kipnis and Shamir proposed a powerful attack against the balanced Oil
+    and Vinegar signature scheme (n = 2v), which finds an equivalent private key in
+    polynomial time. This key can then be used to generate signatures for arbitrary
+    messages.
+
+    INPUT:
+
+    - ``problem`` -- an instance of the UOVProblem class
+    - ``memory_access`` -- specifies the memory access cost model (default: 0, choices: 0 - constant, 1 - logarithmic, 2 - square-root, 3 - cube-root or deploy custom function which takes as input the logarithm of the total memory usage)
+    - ``complexity_type`` -- complexity type to consider (0: estimate, 1: tilde O complexity, default: 0)
+
+    EXAMPLES::
+
+
+    """
+
+    def __init__(self, problem: UOVProblem, **kwargs):
+        super().__init__(problem, **kwargs)
+
+        n, m, _ = self.problem.get_parameters()
+        
+        if n <= 2 * m:
+            raise ValueError('n should be greater than 2 * m')
+
+        self._name = "Kipnis-Shamir"
+        self._gray_code_eval_cost = kwargs.get("gray_code_eval_cost", log2(n))
```

Comment:
`log(q)` as the CollisionAttack?

Reply:
Not used for `kipnis-shamir` it was removed.